posts/go-cli-from-scratch.md

```diff
@@ -0,0 +1,768 @@
+	}
+	os.Exit(0)
+}
+```
+
+Nothing surprising about the `versionFunc`, it prints a message and exits with an all-clear exit code. What might've caught your eye is the ??? for the build and version variables. Those are meant as placeholder variables that will be overwritten during the build process. This allows us to specify version numbers in a config or dynamically include commit hashes on each build. We'll be using ldflags to do this, I'll go over this near the end. For now, if you want to read more check out [using ldflags to set version flags](https://www.digitalocean.com/community/tutorials/using-ldflags-to-set-version-information-for-go-applications).
+
+One point I glanced over is the `func(cmd *Command, args []string)` type for the `Execute` field. That means `Execute` is an anonymous function that can be passed in at any time or extracted into a separate variable as we did for `versionFunc`. Extracting is not necessary but is helpful for building readable code.
+
+The final step would be adding a case statement to our main method but since we already took care of that we're ready for this command.
+
+Now we can accept subcommands and start testing with and without the `--short` flag:
+
+```shell
+$ go run main.go version --short
+brief version: v???
+
+$ go run main.go version
+brief version: v???, build: ???
+```
+
+For now the `???` is fine we'll replace those during the build process later on. At this point we've touched on all the core components of creating a CLI
+
+* Friendly messaging and instructions
+* Adding new subcommands
+* Adding subcommand specific flags
+* Wiring up subcommands together
+
+The next few steps will move faster since we're just repeating the same steps to add more functionality to our CLI.
+
+### Add our Add Command
+
+Next, we'll introduce the functionality to store templates for use. To keep this simple we'll just be keeping a copy of a given template in a specific folder.
+
+As I mentioned we'll just repeat a few similar steps for each new function. I'll start focusing on the repetitive steps and more on the differences. We'll start by defining the usage messaging:
+
+```go
+// add.go
+var addUsage = `Add a template from a file path or URL.
+
+Usage: brief add [OPTIONS] TEMPLATE
+
+Options:
+	--file	path to an existing template file
+`
+```
+
+Next by creating a new add command method which takes a single flag `--file` :
+
+```go
+// add.go
+func NewAddCommand() *Command {
+	cmd := &Command{
+		flags: flag.NewFlagSet("add", flag.ExitOnError),
+		Execute: addFunc,
+	}
+	cmd.flags.StringVar(&filePath, "file", "", "")
+	cmd.flags.Usage = func() {
+		fmt.Fprintln(os.Stderr, addUsage)
+	}
+
+	return cmd
+}
+```
+
+Now we can focus on the functionality. The first thing we need to do is some error handling and validation. We need to confirm the file we're using as a template actually exists. We also need to confirm the directory for saved templates exists, if not we'll go ahead and create it.
```

Comment:
```suggestion
Now we can focus on the functionality. The first thing we need to do is some error handling and validation. We need to confirm the file we're using as a template actually exists. We also need to confirm if the directory for saved templates exists, otherwise, we'll go ahead and create it.
```